File: LeetCode/Study Plan/Practice Problems/Medium/Online Stock Span/Go/online_stock_span.go
package main

type StockSpanner struct{ stack []Pair }

type Pair struct{ val, cnt int }

func Constructor() StockSpanner {
	return StockSpanner{stack: []Pair{}}
}

func (this *StockSpanner) Next(price int) int {
	cnt := 1
	for len(this.stack) > 0 && this.stack[len(this.stack)-1].val <= price {
		cnt += this.stack[len(this.stack)-1].cnt
		this.stack = this.stack[:len(this.stack)-1]
	}
	this.stack = append(this.stack, Pair{price, cnt})
	return cnt
}

File: LeetCode/Medium/Online Stock Span/Java/onlineStockSpan.java
class StockSpanner {
    HashMap<Integer, Integer> spanMap;
    ArrayDeque<Integer> stack;
    public StockSpanner() {
        spanMap = new HashMap<>();
        stack = new ArrayDeque<>();
    }
    
    public int next(int price) {
        if (stack.size() == 0) {
            spanMap.put(price, 1);
            stack.push(price);
        } else {
            int days = 1;
            while(!stack.isEmpty() && stack.peek() <= price) {
                days += spanMap.get(stack.pop());
            }
            spanMap.put(price, days);
            stack.push(price);
        }
        return spanMap.get(price);
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner obj = new StockSpanner();
 * int param_1 = obj.next(price);
 */
